src/query_plan/fetch_dependency_graph.rs

```diff
@@ -292,18 +634,568 @@ impl DeferTracking {
                     selections: Default::default(),
                 })
             }),
-        })
+        }
+    }
+
+    fn update_subselection(
+        &mut self,
+        defer_context: &DeferContext,
+        selection_set: Option<&Arc<NormalizedSelectionSet>>,
+    ) {
+        if !defer_context.is_part_of_query {
+            return;
+        }
+        let Some(primary_selection) = &mut self.primary_selection else {
+            return;
+        };
+        if let Some(parent_ref) = &defer_context.current_defer_ref {
+            self.deferred[parent_ref]
+                .sub_selection
+                .add_at_path(&defer_context.path_to_defer_parent, selection_set)
+        } else {
+            let primary_selection = Arc::make_mut(primary_selection);
+            Arc::make_mut(&mut primary_selection.selections)
+                .add_at_path(&defer_context.path_to_defer_parent, selection_set)
+        }
     }
 }
 
+struct ComputeNodesStackItem<'a> {
+    tree: &'a OpPathTree,
+    node_id: NodeIndex,
+    node_path: FetchDependencyGraphNodePath,
+    context: &'a OpGraphPathContext,
+    defer_context: DeferContext,
+}
+
 pub(crate) fn compute_nodes_for_tree(
+    dependency_graph: &mut FetchDependencyGraph,
+    tree: &OpPathTree,
+    start_node_id: NodeIndex,
+    initial_node_path: FetchDependencyGraphNodePath,
+    initial_defer_context: DeferContext,
+    initial_conditions: &OpGraphPathContext,
+) -> Result<IndexSet<NodeIndex>, FederationError> {
+    let mut stack = vec![ComputeNodesStackItem {
+        tree,
+        node_id: start_node_id,
+        node_path: initial_node_path,
+        context: initial_conditions,
+        defer_context: initial_defer_context,
+    }];
+    let mut created_nodes = IndexSet::new();
+    while let Some(stack_item) = stack.pop() {
+        let node =
+            FetchDependencyGraph::node_weight_mut(&mut dependency_graph.graph, stack_item.node_id)?;
+        for selection_set in &tree.local_selection_sets {
```

Comment:
Indeed this is a bug! Initially I didn’t have a `stack_item` variable but deconstructed it like JS does: `const { tree, group, path, context, deferContext } = stack.pop()!;`, shadowing the previous `tree` binding. I added `stack_item` when splitting this into more functions, because Clippy rightfully complained about taking 12 separate parameters.

I’ll rename the parameter of `compute_nodes_for_tree` to `initial_tree` to avoid this confusion.